cpp/src/join/join_new.cu

```diff
@@ -0,0 +1,223 @@
+
+namespace join {
+
+namespace detail {
+
+  // TODO:
+  //  nested_join_indices      - my function/kernel
+  //  sort_merge_join_indices  - Kumar
+  //  hash_join_indices        - rework current hash mechanism
+  //  filter_join_indices      - new function/kernel
+  //
+  struct nested_loop_join {
+    VectorPair join_indices(table_view const& left,
+                            table_view const& right,
+                            std::vector<join_operation> const& primary_join_ops,
+                            std::vector<join_operation> const& secondary_join_ops,
+                            cudaStream_t stream) {
+
+      return nested_join_indices(left, right, primary_join_ops, stream);
+    }
+  };
+
+  struct sort_merge_join {
+    VectorPair operator()(table_view const& left,
+                          table_view const& right,
+                          std::vector<join_operation> const& primary_join_ops,
+                          std::vector<join_operation> const& secondary_join_ops,
+                          cudaStream_t stream) {
+
+      auto joined_indices = sort_merge_join_indices(left, right, primary_join_ops, stream);
+
+      if (secondary_join_ops.empty()) {
+        return joined_indices;
+      }
+
+      return filter_join_indices(left, right, joined_indices, secondary_join_ops, stream);
+    }
+  };
+
+  struct hash_join {
+    VectorPair operator()(table_view const& left,
+                          table_view const& right,
+                          std::vector<join_operation> const& primary_join_ops,
+                          std::vector<join_operation> const& secondary_join_ops,
+                          cudaStream_t stream) {
+
+      auto joined_indices = hash_join_indices(left, right, primary_join_ops, stream);
+
+      if (secondary_join_ops.empty()) {
+        return joined_indices;
+      }
+
+      return filter_join_indices(left, right, joined_indices, secondary_join_ops, stream);
+    }
+  };
+    
+  bool is_trivial_join(table_view const& left,
+                       table_view const& right,
+                       std::vector<join_operation> const& primary_join_ops,
+                       std::vector<join_operation> const& secondary_join_ops,
+                       join_kind JoinKind) {
+
+    // If there is nothing to join, then send empty table with all columns
+    if (primary_join_ops.empty() && secondary_join_ops.empty()) {
+      return true;
+    }
+
+    // If the inputs are empty, immediately return
+    if ((0 == left.num_rows()) && (0 == right.num_rows())) {
+      return true;
+    }
+
+    // If left join and the left table is empty, return immediately
+    if ((join_kind::LEFT_JOIN == JoinKind) && (0 == left.num_rows())) {
+      return true;
+    }
+
+    // If Inner Join and either table is empty, return immediately
+    if ((join_kind::INNER_JOIN == JoinKind) &&
+        ((0 == left.num_rows()) || (0 == right.num_rows()))) {
+      return true;
+    }
+
+    return false;
+  }
+
```

Comment:
Plan to use the current helper functions defined in join.cu - didn't copy them here (for now).